tests/build_test.sh

```diff
@@ -53,8 +64,189 @@ tearDown()
+{
+  build_config=()
+
+  declare -A expected_build_configurations=(
+    [arch]='x86_64'
+    [cpu_scaling_factor]='100'
+    [kernel_img_name]='bzImage'
+    [menu_config]='nconfig'
+    [doc_type]='htmldocs'
+  )
+
+  parse_configuration "$TMP_DIR/build.config" build_config
+  assertConfigurations build_config expected_build_configurations "$LINENO"
+
+  true # Reset return value
+}
+
+function assertConfigurations()
+{
+  declare -n configurations_ref=$1
+  declare -n expected_configurations_ref=$2
+  local lineno=${3:-LINENO}
+
+  # check if configurations is contained in expected_configurations
+  for k in "${!configurations_ref[@]}"; do
+    if [[ ${expected_configurations_ref[$k]+token} != token ]]; then
+      fail "($lineno): Did not expect setting '$k'."
+    elif [[ ${configurations_ref[$k]} != "${expected_configurations_ref[$k]}" ]]; then
+      fail "($lineno): Expected setting '${k}' to be '${expected_configurations_ref[$k]}' (found '${configurations_ref[$k]}')."
+    fi
+  done
+
+  # check if configurations has all expected_configurations keys
+  for k in "${!expected_configurations_ref[@]}"; do
+    if [[ ${configurations_ref[$k]+token} != token ]]; then
+      fail "($lineno): Expected setting '$k' to be present."
+    fi
+  done
+}
+
+# To test the order of config file loading, we will put a file named
+# kworkflow.config in each place, in order, and remove the previous one.
+# The order is: PWD, XDG_CONFIG_HOME, XDG_CONFIG_DIRS, KW_ETC_DIR
+function test_parse_configuration_files_loading_order()
+{
+  build_config=()
+
+  local expected
+  local original_dir="$PWD"
+
+  KW_ETC_DIR='1'
+  XDG_CONFIG_DIRS='2:3:4'
+  XDG_CONFIG_HOME='5'
+
+  expected=(
+    "1/$BUILD_CONFIG_FILENAME"
+    "4/$KWORKFLOW/$BUILD_CONFIG_FILENAME"
+    "3/$KWORKFLOW/$BUILD_CONFIG_FILENAME"
+    "2/$KWORKFLOW/$BUILD_CONFIG_FILENAME"
+    "5/$KWORKFLOW/$BUILD_CONFIG_FILENAME"
+  )
+
+  output="$(
+    function parse_configuration()
+    {
+      printf '%s\n' "$1"
+    }
+    load_build_config
+  )"
+
+  compare_command_sequence 'Wrong config file reading order' "$LINENO" 'expected' "$output"
+
+  # IF XDG global variables are not defined
+  unset XDG_CONFIG_DIRS
+  unset XDG_CONFIG_HOME
+  HOME='5'
+
+  expected=(
+    "1/$BUILD_CONFIG_FILENAME"
+    "/etc/xdg/$KWORKFLOW/$BUILD_CONFIG_FILENAME"
+    "5/.config/$KWORKFLOW/$BUILD_CONFIG_FILENAME"
+    "$PWD/$BUILD_CONFIG_FILENAME"
+  )
+
+  output="$(
+    function parse_configuration()
+    {
+      printf '%s\n' "$1"
+    }
+    load_build_config
+  )"
+
+  compare_command_sequence 'Wrong config file reading order' "$LINENO" 'expected' "$output"
+}
+
+function test_show_build_variables_completeness()
+{
+  build_config=()
+
+  local -A shown_options
+  local -A possible_options
+  local output
+
+  # get all assigned options, including commented ones
+  # remove #'s and ='s to get option names
+  output="$(cat "$original_dir/etc/build_template.config")"
```

Comment:
You can use this optimization:
https://kworkflow.org/content/codingstyle.html#command-substitution-and-arithmetic-expression